src/metax_api/services/file_service.py

```diff
@@ -779,64 +798,57 @@ def _get_directory_contents(cls, directory_id, recursive=False, max_depth=1, dep
         """
 
         # select dirs which are contained by the directory,
         # AND which contain files belonging to the cr <-> files m2m relation table,
-        # AND there exists files for CR which beging with the same path as the dir path,
+        # AND there exists files for CR which begin with the same path as the dir path,
         # to successfully also include files which did not DIRECTLY contain any files, but do
         # contain files further down the tree.
-        #
         # dirs which otherwise contained files, but not any files that were selected for
         # the cr, are not returned.
+
         sql_select_dirs_for_cr = """
             select d.id
             from metax_api_directory d
-            where d.parent_directory_id = %s
+            where d.parent_directory_id = {}
             and exists(
                 select 1
                 from metax_api_file f
                 inner join metax_api_catalogrecord_files cr_f on cr_f.file_id = f.id
                 where f.file_path like (d.directory_path || '/%%')
-                and cr_f.catalogrecord_id = %s
+                and cr_f.catalogrecord_id {} {}
                 and f.removed = false
                 and f.active = true
             )
             """
 
-        # select files which are contained by the directory, and which
-        # belong to the cr <-> files m2m relation table
-        sql_select_files_for_cr = """
-            select f.id
-            from metax_api_file f
-            inner join metax_api_catalogrecord_files cr_f on cr_f.file_id = f.id
-            where f.parent_directory_id = %s
-            and cr_f.catalogrecord_id = %s
-            and f.removed = false
-            and f.active = true
-            """
-
         with connection.cursor() as cr:
-            cr.execute(sql_select_dirs_for_cr, [directory_id, cr_id])
-            directory_ids = [ row[0] for row in cr.fetchall() ]
-
             if dirs_only:
-                file_ids = []
+                files = []
             else:
-                cr.execute(sql_select_files_for_cr, [directory_id, cr_id])
-                file_ids = [ row[0] for row in cr.fetchall() ]
+                if cr_id:
+                    cr.execute(sql_select_dirs_for_cr.format(directory_id, '=', cr_id))
+                    files = None if dirs_only else File.objects \
+                        .filter(record__pk=cr_id, parent_directory=directory_id, removed=False, active=True) \
+                        .values(*file_fields)
+                if not_cr_id:
+                    cr.execute(sql_select_dirs_for_cr.format(directory_id, '!=', not_cr_id))
+                    files = None if dirs_only else File.objects.exclude(record__pk=not_cr_id) \
+                        .filter(parent_directory=directory_id, removed=False, active=True) \
+                        .values(*file_fields)
+            directory_ids = [ row[0] for row in cr.fetchall() ]
```

Comment:
> The if dirs_only on line 831 seems obsolete now I think?
Yep

> The if cr_id and if not_cr_id should be probably be written as elif, since they should not be simultaneously? I noticed an error is already raised elsewhere about that, but it would make code more explicit here too.
Yep

> In general about using ORM in Metax, when using the objects manager, e.g. File.objects.xyz, the removed and active filters do not have to be separetely defined, as they have already been defined in a superclass in models/common.py. When needing to query from all objects in the db, use the objects_unfiltered manager instead, which does not have any pre-set filters.
This needs more explanation :)

> cr.execute(sql_select_dirs_for_cr.format(directory_id, '=', cr_id)) SQL prepared statements have been replaced with string format(), which while in this case should be ok since all the values should be able to be trusted, but it opens the door for future bugs when changes occur. I would recommend keeping the use of prepared statements for directory_id and cr_id, while the hardcoded '=' can be added in a previous step using format().
This needs more explanation too :)